File: prd-api/src/PrdAgent.Core/Services/ArticleMarkerExtractor.cs
using System.Text.RegularExpressions;
using PrdAgent.Core.Models;

namespace PrdAgent.Core.Services;

/// <summary>
/// 文章配图标记提取器：提取和替换 [[...]] 标记
/// </summary>
public static class ArticleMarkerExtractor
{
    private static readonly Regex MarkerRegex = new(@"\[插图\]\s*:\s*(.+?)(?=\n|$)", RegexOptions.Compiled | RegexOptions.Multiline);

    /// <summary>
    /// 从文章内容中提取所有 [[...]] 标记
    /// </summary>
    public static List<ArticleMarker> Extract(string content)
    {
        var markers = new List<ArticleMarker>();
        var matches = MarkerRegex.Matches(content ?? string.Empty);

        for (int i = 0; i < matches.Count; i++)
        {
            var match = matches[i];
            markers.Add(new ArticleMarker
            {
                Index = i,
                Text = match.Groups[1].Value,
                StartPos = match.Index,
                EndPos = match.Index + match.Length
            });
        }

        return markers;
    }

    /// <summary>
    /// 将文章中的 [[...]] 标记替换为实际图片 Markdown 链接
    /// </summary>
    public static string ReplaceMarkersWithImages(string content, List<ImageAsset> assets)
    {
        var sortedAssets = assets
            .Where(a => a.ArticleInsertionIndex.HasValue)
            .OrderBy(a => a.ArticleInsertionIndex)
            .ToList();

        var result = content;
        var offset = 0;

        foreach (var asset in sortedAssets)
        {
            var marker = MarkerRegex.Match(result, offset);
            if (!marker.Success) break;

            var altText = asset.OriginalMarkerText ?? "配图";
            var replacement = $"![{altText}]({asset.Url})";
            result = result.Remove(marker.Index, marker.Length)
                          .Insert(marker.Index, replacement);
            offset = marker.Index + replacement.Length;
        }

        return result;
    }
}

/// <summary>
/// 文章配图标记数据模型
/// </summary>
public class ArticleMarker
{
    public int Index { get; set; }
    public string Text { get; set; } = string.Empty;
    public int StartPos { get; set; }
    public int EndPos { get; set; }
}

File: prd-admin/src/lib/articleMarkerExtractor.ts
import React from 'react';

export type ArticleMarker = {
  index: number;
  text: string;
  startPos: number;
  endPos: number;
};

/**
 * 从文章内容中提取所有 [插图] : ... 标记
 */
export function extractMarkers(content: string): ArticleMarker[] {
  const regex = /\[插图\]\s*:\s*(.+?)(?=\n|$)/gm;
  const markers: ArticleMarker[] = [];
  let match;
  let index = 0;

  while ((match = regex.exec(content)) !== null) {
    const text = match[1].trim();
    // 跳过空标记（流式输出时可能只有 [插图] : 但还没有描述内容）
    if (text.length === 0) {
      continue;
    }
    markers.push({
      index: index++,
      text,
      startPos: match.index,
      endPos: match.index + match[0].length,
    });
  }

  return markers;
}

/**
 * 高亮显示文章中的 [插图] : ... 标记
 */
export function highlightMarkers(content: string): React.ReactNode {
  const parts: React.ReactNode[] = [];
  let lastIndex = 0;
  const regex = /\[插图\]\s*:\s*(.+?)(?=\n|$)/gm;
  let match;

  while ((match = regex.exec(content)) !== null) {
    // 普通文本
    if (match.index > lastIndex) {
      parts.push(content.slice(lastIndex, match.index));
    }
    // 高亮标记
    parts.push(
      React.createElement(
        'mark',
        {
          key: match.index,
          className: 'bg-yellow-200 dark:bg-yellow-800 px-1 rounded',
        },
        `[插图] : ${match[1].trim()}`
      )
    );
    lastIndex = match.index + match[0].length;
  }

  // 剩余文本
  if (lastIndex < content.length) {
    parts.push(content.slice(lastIndex));
  }

  return React.createElement(React.Fragment, null, ...parts);
}

/**
 * 替换文章中的 [插图] : ... 标记为图片 Markdown 链接
 */
export function replaceMarkersWithImages(
  content: string,
  images: Array<{ url: string; alt?: string }>
): string {
  const regex = /\[插图\]\s*:\s*(.+?)(?=\n|$)/gm;
  let result = content;
  let offset = 0;

  images.forEach((img) => {
    const match = regex.exec(content);
    if (!match) return;

    const replacement = `![${img.alt || match[1].trim()}](${img.url})`;
    const startPos = match.index + offset;
    const endPos = startPos + match[0].length;

    result = result.slice(0, startPos) + replacement + result.slice(endPos);
    offset += replacement.length - match[0].length;
  });

  return result;
}
